src/lang/base/Bech32.ml

```diff
@@ -0,0 +1,189 @@
+let bech32_polymod_step pre  =
+    let b = pre lsr 25 in
+    ((pre land 0x1FFFFFF) lsl 5) lxor
+        (-((b lsr 0) land 1) land 0x3b6a57b2) lxor
+        (-((b lsr 1) land 1) land 0x26508e6d) lxor
+        (-((b lsr 2) land 1) land 0x1ea119fa) lxor
+        (-((b lsr 3) land 1) land 0x3d4233dd) lxor
+        (-((b lsr 4) land 1) land 0x2a1462b3)
+
+let ascii_of_char c = Caml.Char.code c
+let char_of_ascii i = Caml.Char.chr i
+let bytes_of_bitstring bs = Bytes.of_string @@ string_of_bitstring bs
+
+(* Decodes a bech32 address string to a string of 20 bytes.
+ * Signature: prefix -> bech32_addr -> bystr20 address option. *)
+let decode_bech32_addr ~prefix ~addr:str =
+  (* 1. Must be of the right length. *)
+  if String.length str <> (bech32_addr_len ~prefix) then None else
+  (* 2. Must begin with prefix. *)
+  if (String.sub str ~pos:0 ~len:(String.length prefix)) <> prefix then None else
+  (* 3. Must have separator "1" after prefix. *)
+  if (String.sub str ~pos:(String.length prefix) ~len:1) <> "1" then None else
+
+  (* 4. Scan the prefix for errors. *)
+  let (chk, err, (have_lower, have_upper)) = 
+    List.fold (String.to_list prefix)  ~init:(1, false, (false, false)) ~f:(fun (chk, err, (have_lower, have_upper)) ch ->
+      if err || (ascii_of_char ch) < 33 || (ascii_of_char ch) > 126 then (chk, true, (have_lower, have_upper)) else
+ 
+      let (ch', have_lower', have_upper') =
+        if (ascii_of_char ch) >= (ascii_of_char 'a') && (ascii_of_char ch) <= (ascii_of_char 'z') then 
+          (ascii_of_char ch, true, have_upper)
+        else if (ascii_of_char ch) >= (ascii_of_char 'A') && (ascii_of_char ch) <= (ascii_of_char 'Z') then
+          ((ascii_of_char ch) - (ascii_of_char 'A') + (ascii_of_char 'a'), have_lower, true)
+        else
+          (ascii_of_char ch, have_lower, have_upper)
+      in
+      let chk' = (bech32_polymod_step chk) lxor (ch' lsr 5) in
+      (chk', err, (have_lower', have_upper'))
+    )
+  in
+  if err || (have_lower && have_upper) then None else
+
+  (* 5. Checksum the prefix *)
+  let chk = bech32_polymod_step chk in
+  let chk = List.fold (String.to_list prefix) ~init:chk ~f:(fun acc_chk c ->
+    (bech32_polymod_step acc_chk) lxor ((ascii_of_char c) land 0x1f)
+  ) in
+
+  (* 6. Scan the addr and checksum for errors. *)
+  let addr_chk_str = String.sub str ~pos:((String.length prefix)+1) ~len:(32+6) in
+  (* Create a buffer of bits for the result (20-byte raw address). *)
+  let bitacc = Buffer.create () in
+  let (chk, err, (have_lower, have_upper), _) = List.fold (String.to_list addr_chk_str) ~init:(chk, false, (false, false), 0)
+   ~f:(fun (chk, err, (have_lower, have_upper), index) c ->
+    (* do nothing if we've already seen an error. *)
+    if err then (chk, err, (have_lower, have_upper), index+1) else
+
+    let c' =
+      (* it's an error if any bit other than the least 7 bits are 1. *)
+      if (ascii_of_char c) land (Int.minus_one lsl 7) <> 0 then -1 else
```

Comment:
The idea behind this check is that, only values that are in the range 0-127 are valid. In the case of the C code, since the data type is `uint8`, it's good enough for them to check that the last bit is not `1`. I instead check that all bits but the least 7 are 0s. So they are equivalent.

Reply:
Ah, that makes sense.